### PR TITLE
Send 204 status code instead of 200 to Twitch when handling notifications

### PR DESCRIPTION
**Pull request by Satont**
Otherwise twitch will retry requests

### CODE REVIEW

**Comment by dnsge** [comment]
> Otherwise twitch will retry requests

Are you sure this is the case? According to https://dev.twitch.tv/docs/eventsub/handling-webhook-events/#processing-an-event, 

> Your response should return a 2XX status code

Which doesn't demand anything about the body content. In the full example at the bottom of that page, the handler only returns a `204` status code with an empty response. I've been using this library in production for quite a while now and haven't run into any retrying issues.

**Comment by Satont** [comment]
> > Otherwise twitch will retry requests
> 
> Are you sure this is the case? According to https://dev.twitch.tv/docs/eventsub/handling-webhook-events/#processing-an-event,
> 
> > Your response should return a 2XX status code
> 
> Which doesn't demand anything about the body content. In the full example at the bottom of that page, the handler only returns a `204` status code with an empty response. I've been using this library in production for quite a while now and haven't run into any retrying issues.

Yes, i tested it. I recieved events 3 times with `retry` 1,2,3. When i added this `Write`, i got it only once.

**Comment by dnsge** [comment]
I've talked with a few other people that use EventSub Webhooks and they concur that `w.WriteStatus(http.StatusOK)` should be enough here. Could you provide a minimal example to reproduce the retrying?

Another idea is that perhaps you are getting duplicate notifications from Twitch (though you say "retry" so I doubt it). Are you using an `IDTracker` in your `SubHandler`? (I know there are no docs for this repo so you may not have known about that feature)

**Comment by Satont** [comment]
> I've talked with a few other people that use EventSub Webhooks and they concur that w.WriteStatus(http.StatusOK) should be enough here. Could you provide a minimal example to reproduce the retrying?

```go
manager := eventsub_framework.NewSubHandler(true, []byte(services.Config.TwitchClientSecret))
manager.HandleChannelUpdate = func(h *eventsub_bindings.ResponseHeaders, event *eventsub_bindings.EventChannelUpdate) {
  fmt.Prinln(h.MessageRetry)
}
  
  go func() {
		if err := http.Serve(appTun, manager); err != nil {
			panic(err)
		}
	}()
```

Thats bare minimum i used. So, i got 3 events, then twitch stop retrying because they assume we are down because no "OK" response.

I asked in Twitch Libs discord dev, @d-fischer said, we should send `http.StatusNoContent` with no body or `http.StatusOk` with any body.

```
d-fischer / SatisfiedPear — Yesterday at 11:58 PM
That's technically against the HTTP spec
Either send 204, or send some body.
```

> Another idea is that perhaps you are getting duplicate notifications from Twitch (though you say "retry" so I doubt it). Are you using an IDTracker in your SubHandler? (I know there are no docs for this repo so you may not have known about that feature)

No, im not using `IDTracker`, but i saw that in source.

**Comment by dnsge** [comment]
I took your minimal example and made it slightly less minimal so it would print out a bunch of extra info and also create the EventSub subscription at run. If you look through the output below, you'll see that when I change the title of my stream, I receive a single event with `MessageRetry:0`. 

I'm leaning to this being an issue with your proxy (or rather, how your proxy handles an HTTP response that is probably not perfectly following the spec).

Code:
```go

func main() {
	handler := esf.NewSubHandler(true, []byte(signingKey))
	handler.HandleChannelUpdate = func(h *esb.ResponseHeaders, event *esb.EventChannelUpdate) {
		pretty.Println(h)
		pretty.Println(event)
	}
	handler.VerifyChallenge = func(h *esb.ResponseHeaders, chal *esb.SubscriptionChallenge) bool {
		pretty.Println(h)
		pretty.Println(chal)
		return true
	}

	var wg sync.WaitGroup

	wg.Add(1)
	go func() {
		log.Println("Listening at 127.0.0.1:8080")
		err := http.ListenAndServe("127.0.0.1:8080", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			pretty.Println(r.Header)
			handler.ServeHTTP(w, r)
		}))
		if err != nil {
			fmt.Println(err)
		}
	}()

	wg.Add(1)
	go func() {
		cli := esf.NewSubClient(&staticCreds{})
		_, _ = cli.Subscribe(context.Background(), &esf.SubRequest{
			Type: "channel.update",
			Condition: esb.ConditionChannelUpdate{
				BroadcasterUserID: "132532813",
			},
			Callback: "https://my_proxy/webhooks",
			Secret:   signingKey,
		})
	}()

	wg.Wait()
}

type staticCreds struct{}

func (c *staticCreds) ClientID() (string, error) {
	return clientID, nil
}

func (c *staticCreds) AppToken() (string, error) {
	return appToken, nil
}

```

Output:
```
2023/03/05 18:02:49 Listening at 127.0.0.1:8080
http.Header{
    "Accept-Encoding":                                  {"gzip"},
    "Connection":                                       {"close", "close"},
    "Content-Length":                                   {"399"},
    "Content-Type":                                     {"application/json"},
    "Twitch-Eventsub-Message-Id":                       {"0833f895-a8b0-4f27-b22c-4e065c061a0f"},
    "Twitch-Eventsub-Message-Retry":                    {"0"},
    "Twitch-Eventsub-Message-Signature":                {"sha256=f21dcf94ec4bca3b03351e33eb5e2d8fd32f997329622a84fd68fcffe5b81b85"},
    "Twitch-Eventsub-Message-Timestamp":                {"2023-03-05T23:02:49.244171299Z"},
    "Twitch-Eventsub-Message-Type":                     {"webhook_callback_verification"},
    "Twitch-Eventsub-Subscription-Is-Batching-Enabled": {"false"},
    "Twitch-Eventsub-Subscription-Type":                {"channel.update"},
    "Twitch-Eventsub-Subscription-Version":             {"1"},
    "User-Agent":                                       {"Go-http-client/1.1"},
    "X-Forwarded-For":                                  {"XXX"},
    "X-Forwarded-Proto":                                {"https"},
    "X-Real-Ip":                                        {"XXX"},
}
&eventsub_bindings.ResponseHeaders{MessageID:"0833f895-a8b0-4f27-b22c-4e065c061a0f", MessageRetry:0, MessageType:"webhook_callback_verification", MessageSignature:"sha256=f21dcf94ec4bca3b03351e33eb5e2d8fd32f997329622a84fd68fcffe5b81b85", MessageTimestamp:"2023-03-05T23:02:49.244171299Z", SubscriptionType:"channel.update", SubscriptionVersion:"1"}
&eventsub_bindings.SubscriptionChallenge{
    Challenge:    "S2mjgiyVcIMpuSzlFrrG6552OJvjpkxpXCDEPxqIA-Y",
    Subscription: eventsub_bindings.Subscription{
        ID:        "2d51bb7e-793f-49c6-9a30-ea329a7f5d20",
        Type:      "channel.update",
        Version:   "1",
        Status:    "webhook_callback_verification_pending",
        Cost:      0,
        Condition: map[string]interface {}{
            "broadcaster_user_id": "132532813",
        },
        CreatedAt: "2023-03-05T23:02:49.237093498Z",
    },
}

... some time later ...

http.Header{
    "Accept-Encoding":                                  {"gzip"},
    "Connection":                                       {"close", "close"},
    "Content-Length":                                   {"538"},
    "Content-Type":                                     {"application/json"},
    "Twitch-Eventsub-Message-Id":                       {"shUq7LSg_ys6-KNbiSxdIHpxIGj-dvWL0w4gIt970QI="},
    "Twitch-Eventsub-Message-Retry":                    {"0"},
    "Twitch-Eventsub-Message-Signature":                {"sha256=72f036a1ad195a7a57fb1f114f52adebd7249f65265c97ce384e02d7f06230de"},
    "Twitch-Eventsub-Message-Timestamp":                {"2023-03-05T23:03:10.72000433Z"},
    "Twitch-Eventsub-Message-Type":                     {"notification"},
    "Twitch-Eventsub-Subscription-Is-Batching-Enabled": {"false"},
    "Twitch-Eventsub-Subscription-Type":                {"channel.update"},
    "Twitch-Eventsub-Subscription-Version":             {"1"},
    "User-Agent":                                       {"Go-http-client/1.1"},
    "X-Forwarded-For":                                  {"XXX"},
    "X-Forwarded-Proto":                                {"https"},
    "X-Real-Ip":                                        {"XXX"},
}
&eventsub_bindings.ResponseHeaders{MessageID:"shUq7LSg_ys6-KNbiSxdIHpxIGj-dvWL0w4gIt970QI=", MessageRetry:0, MessageType:"notification", MessageSignature:"sha256=72f036a1ad195a7a57fb1f114f52adebd7249f65265c97ce384e02d7f06230de", MessageTimestamp:"2023-03-05T23:03:10.72000433Z", SubscriptionType:"channel.update", SubscriptionVersion:"1"}
&eventsub_bindings.EventChannelUpdate{BroadcasterUserID:"132532813", BroadcasterUserLogin:"icelys", BroadcasterUserName:"icelys", Title:"hello there! 123", Language:"en", CategoryID:"509658", CategoryName:"Just Chatting", IsMature:false}
```

**Comment by dnsge** [comment]
I think that my prefered solution to this would be to replace the `w.WriteHeader(http.StatusOK)` call with

```go
w.Header().Set("Content-Length", "0")
w.WriteHeader(http.StatusNoContent)
```

instead

File: handler.go [hunk 1]
```diff
@@ -490,6 +490,7 @@ func (s *SubHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			}
 
 			w.WriteHeader(http.StatusOK)
```
Comment (dnsge) [inline handler.go]:
```suggestion
			w.Header().Set("Content-Length", "0")
```

File: handler.go [hunk 2]
```diff
@@ -490,6 +490,7 @@ func (s *SubHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+			w.Write([]byte("OK"))
```
Comment (dnsge) [inline handler.go]:
```suggestion
			w.WriteHeader(http.StatusNoContent)
```

**Review by dnsge — CHANGES_REQUESTED** [review]

**Comment by Satont** [comment]
Done.

**Comment by dnsge** [comment]
Great, thank you for noticing this problem!

**Comment by Satont** [comment]
Can you release new tag please?

**Comment by dnsge** [comment]
Done, thanks for the reminder

**Comment by Satont** [comment]
Thanks for the lib. You simplified my project a bit. :)

**Pull request merged**